cpp/src/arrow/compute/kernels/pivot_internal.cc

```diff
@@ -18,110 +18,137 @@
+    int64_t total_length = 0;
+    for (const auto& key : options->key_names) {
+      total_length += static_cast<int64_t>(key.length());
+    }
+    RETURN_NOT_OK(builder.ReserveData(total_length));
+    for (const auto& key : options->key_names) {
+      builder.UnsafeAppend(key);
+    }
+    ARROW_ASSIGN_OR_RAISE(auto binary_key_array, builder.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto key_array,
+                          Cast(*binary_key_array, &key_type, CastOptions::Safe(), ctx));
+    // Populate the grouper with the keys from the array
+    ExecSpan batch({ExecValue(*key_array->data())}, key_array->length());
+    RETURN_NOT_OK(grouper_->Populate(batch));
+    if (grouper_->num_groups() != options->key_names.size()) {
+      // There's a duplicate key, find it to emit a nicer error message
+      std::unordered_set<std::string_view> seen;
+      for (const auto& key : options->key_names) {
+        auto [_, inserted] = seen.emplace(key);
+        if (!inserted) {
+          return Status::KeyError("Duplicate key name '", key, "' in PivotWiderOptions");
+        }
       }
+      Unreachable("Grouper doesn't agree with std::unordered_set");
     }
-    unexpected_key_behavior_ = options->unexpected_key_behavior;
     return Status::OK();
   }
 
- protected:
-  Result<PivotWiderKeyIndex> KeyNotFound(std::string_view key_name) {
-    if (unexpected_key_behavior_ == PivotWiderOptions::kIgnore) {
-      return kNullPivotKey;
+  Result<std::shared_ptr<ArrayData>> MapKeys(const ArraySpan& array) override {
+    if (array.GetNullCount() != 0) {
+      return NullKeyName();
     }
-    DCHECK_EQ(unexpected_key_behavior_, PivotWiderOptions::kRaise);
-    return Status::KeyError("Unexpected pivot key: ", key_name);
+    return MapKeysInternal(array, array.length);
   }
 
-  Result<PivotWiderKeyIndex> LookupKey(std::string_view key_name) {
-    const auto it = this->key_name_map_.find(key_name);
-    if (ARROW_PREDICT_FALSE(it == this->key_name_map_.end())) {
-      return KeyNotFound(key_name);
+  Result<std::optional<PivotWiderKeyIndex>> MapKey(const Scalar& scalar) override {
+    if (!scalar.is_valid) {
+      return NullKeyName();
+    }
+    ARROW_ASSIGN_OR_RAISE(auto group_id_array, MapKeysInternal(&scalar, /*length=*/1));
+    DCHECK_EQ(group_id_array->length, 1);
+    if (group_id_array->GetNullCount() == 0) {
+      return group_id_array->GetValues<uint32_t>(1)[0];
     } else {
-      return it->second;
+      return std::nullopt;
+    }
+  }
+
+ protected:
+  Result<std::shared_ptr<ArrayData>> MapKeysInternal(const ExecValue& values,
+                                                     int64_t length) {
+    ARROW_ASSIGN_OR_RAISE(auto result, grouper_->Lookup(ExecSpan({values}, length)));
+    DCHECK(result.is_array());
+    DCHECK_EQ(result.type()->id(), Type::UINT32);
+    auto group_id_array = result.array();
+    if (ARROW_PREDICT_FALSE(group_id_array->GetNullCount() != 0) &&
```

Comment:
Maybe extract `group_id_array->GetNullCount() != 0` as `has_nulls` variable for better readability? `ARROW_PREDICT_FALSE(.. != 0)` is slightly harder to interpret.

Reply:
Ok, will do.